### PR TITLE
Etl rewritten and simplified

### PR DESCRIPTION
**Pull request by mrmasterplan**
I completely re-thought the whole ETL library. 
**There are backwards incompatible changes** hence the move to version 0.4.
There are no more special cases for multiple vs single datasets on input or output side.
The documentation has been updated to reflect all the simplified changes.
All `Multiple...` and `Delegating...` classes are gone! Have a look at how little code there is left in the `atc.etl` library!
The code can do everything that it could before plus:
- it is now possible to specify the key for the dataset returned by an extractor (for use in multi-transform)
- it is now possible to reference the result of previous extraction steps in case of complex extraction logic (we have seen examples in practice)
- `save_many` now exists although I see no usecase for it.
- All orchestration is now streamlined so any sequence of transform, load, extract steps can be orchestrated easily.
- the examples are now executed as part of unit testing. They can then be inserted into the readme by a simple script.

### CODE REVIEW

File: docs/etl/insert_examples.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from pathlib import Path
```
Comment (christianhelle) [inline docs/etl/insert_examples.py:1]: This is such a cool way of building docs!

File: src/atc/etl/loader.py [hunk 2]
```diff
@@ -1,24 +1,31 @@
...
 from pyspark.sql import DataFrame
 
+from atc.etl.types import dataset_group, EtlBase
 
-class Loader:
-    @abstractmethod
-    def save(self, df: DataFrame) -> DataFrame:
-        return df
 
+class Loader(EtlBase):
```
Comment (christianhelle) [inline src/atc/etl/loader.py:8]: I agree that a `Loader` is by design a data sink. The `DataFrame` return type was something I added previously for `Loader` implementations that target CosmosDb, where the document ID and partition key are required by the data store but not relevant on the data record. Returning the `DataFrame` with the new columns made sense at the time. In a way it makes more sense that its the `Transformer` that appends these columns but then the `Transformer` is now responsible for data storage specific concerns, and not purely data transformation

File: src/atc/etl/orchestrator.py [hunk 3]
```diff
@@ -1,142 +1,35 @@
...
-class SingleExtractorOrchestrator(Orchestrator):
-    def __init__(self,
-        extractor: Extractor,
-        transformer: Union[Transformer, DelegatingTransformer],
-        loader: Union[Loader, DelegatingLoader]
-    ):
-        self.extractor = extractor
-        self.transformer = transformer
-        self.loader = loader
+    """
```
Comment (christianhelle) [inline src/atc/etl/orchestrator.py:9]: I love this!

File: tests/etl/test_orchestrator.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import subprocess
+import sys
+import unittest
+from pathlib import Path
+
+
+class OrchestratorTests(unittest.TestCase):
+    def test_examples(self):
```
Comment (christianhelle) [inline tests/etl/test_orchestrator.py:8]: How does this work? Will it the test runner understand that the sub-processes are unit tests? and will it propagate individual test results back to the top level process?

Comment (christianhelle) [reply tests/etl/test_orchestrator.py:8]: I love the fact that this test enforces us to keep the example code alive and up to date

**Review by christianhelle — COMMENTED** [review]
These changes are really cool!

File: src/atc/etl/orchestrator.py [hunk 5]
```diff
@@ -1,142 +1,35 @@
...
+    It is up to the user of this library that extractors,
```
Comment (christianhelle) [inline src/atc/etl/orchestrator.py:10]: I love this!

File: src/atc/etl/orchestrator.py (hunk 3 above)
Comment (christianhelle) [inline src/atc/etl/orchestrator.py:9]: >>    It is up to the user of this library that extractors,
>>   transformers and loaders live up to their names and are not
>>    used in a wrong order.

I love this!

File: tests/etl/test_orchestrator.py (hunk 4 above)
Comment (mrmasterplan) [reply tests/etl/test_orchestrator.py:8]: This simply validates that the examples can be executed. Check=True checks that the process exits normally without exception (no error code).

File: src/atc/etl/loader.py (hunk 2 above)
Comment (mrmasterplan) [reply src/atc/etl/loader.py:8]: I am familiar with the use case, and it is fine to both add it in the loader or in a transformer. The real reason to not return it here is to make it clear that in te case of loader->loader->loader, the loaders will be called with the same dataframe and not sequentially enriched dataframes.

**Comment by mrmasterplan** [comment]
@JeppeBlixen I would really appreciate your review since I think you use this in production, and some changes are non-compatible. I recently moved to fixed dependencies, so I know my system won't break, but yours might.

File: examples/etl/basic_example.py [hunk 6]
```diff
@@ -2,51 +2,54 @@
...
         df.printSchema()
         return df
 
 
 class NoopLoader(Loader):
-    def save(self, df: DataFrame) -> DataFrame:
-        df.write.format('noop').mode('overwrite').save()
-        return df
+    def save(self, df: DataFrame) -> None:
+        df.write.format("noop").mode("overwrite").save()
```
Comment (LauJohansson) [inline examples/etl/basic_example.py:43]: why not returning the df?

File: examples/etl/multiple_output_example.py [hunk 7]
```diff
@@ -2,58 +2,62 @@
...
         df.printSchema()
         return df
 
 
 class NoopSilverLoader(Loader):
-    def save(self, df: DataFrame) -> DataFrame:
-        df.write.format('noop').mode('overwrite').save()
-        return df
+    def save(self, df: DataFrame) -> None:
+        df.write.format("noop").mode("overwrite").save()
```
Comment (LauJohansson) [inline examples/etl/multiple_output_example.py:45]: why not returning the df?

File: examples/etl/multiple_transform_example.py [hunk 8]
```diff
@@ -32,18 +37,18 @@ def process(self, df: DataFrame) -> DataFrame:
...
-print('ETL Orchestrator using multiple transformers')
-etl = (Orchestration
-       .extract_from(GuitarExtractor())
-       .transform_with(IntegerColumnTransformer('id'))
-       .transform_with(IntegerColumnTransformer('year'))
-       .load_into(NoopLoader())
-       .build())
+    def save(self, df: DataFrame) -> None:
+        df.write.format("noop").mode("overwrite").save()
+
```
Comment (LauJohansson) [inline examples/etl/multiple_transform_example.py:42]: why not returning the df?

File: examples/etl/no_transform_example.py [hunk 9]
```diff
@@ -1,37 +1,38 @@
...
+            ),
+        )
 
 
 class NoopLoader(Loader):
-    def save(self, df: DataFrame) -> DataFrame:
-        df.write.format('noop').mode('overwrite').save()
-        return df
+    def save(self, df: DataFrame) -> None:
+        df.write.format("noop").mode("overwrite").save()
```
Comment (LauJohansson) [inline examples/etl/no_transform_example.py:31]: return df?

File: examples/etl/multiple_input_example.py [hunk 10]
```diff
@@ -2,60 +2,70 @@
...
+            ),
+        )
+
+
+class CountryOfOriginTransformer(Transformer):
     def process_many(self, dataset: {}) -> DataFrame:
-        usa_df = dataset['AmericanGuitarExtractor'].withColumn('country', f.lit('USA'))
-        jap_df = dataset['JapaneseGuitarExtractor'].withColumn('country', f.lit('Japan'))
+        usa_df = dataset["AmericanGuitarExtractor"].withColumn("country", f.lit("USA"))
+        jap_df = dataset["japanese"].withColumn("country", f.lit("Japan"))
```
Comment (LauJohansson) [inline examples/etl/multiple_input_example.py:52]: Just for extra niceness. Consider renaming the dataset keys to be more similar.

Comment (LauJohansson) [reply examples/etl/multiple_input_example.py:52]: e.g. "AmericanGuitars" and "JapaneseGuitars"

File: examples/etl/multiple_input_example.py [hunk 11]
```diff
@@ -2,60 +2,70 @@
...
+        df.write.format("noop").mode("overwrite").save()
+
+
+print("ETL Orchestrator using multiple extractors")
+etl = (
+    Orchestrator()
+    .extract_from(AmericanGuitarExtractor())
+    .extract_from(JapaneseGuitarExtractor())
+    .transform_with(CountryOfOriginTransformer())
+    .load_into(NoopLoader())
```
Comment (LauJohansson) [inline examples/etl/multiple_input_example.py:67]: great idea to remove the ".build()" method!

File: examples/etl/multiple_output_example.py [hunk 12]
```diff
@@ -2,58 +2,62 @@
...
-
-print('ETL Orchestrator using multiple loaders')
-etl = (Orchestration
-       .extract_from(GuitarExtractor())
-       .transform_with(BasicTransformer())
-       .load_into(NoopSilverLoader())
-       .load_into(NoopGoldLoader())
-       .build())
+    def save(self, df: DataFrame) -> None:
+        df.write.format("noop").mode("overwrite").save()
```
Comment (LauJohansson) [inline examples/etl/multiple_output_example.py:50]: return the df?

File: examples/etl/basic_example.py (hunk 6 above)
Comment (mrmasterplan) [reply examples/etl/basic_example.py:43]: https://github.com/atc-net/atc-dataplatform/pull/57#discussion_r780823779

File: examples/etl/multiple_output_example.py (hunk 7 above)
Comment (mrmasterplan) [reply examples/etl/multiple_output_example.py:45]: https://github.com/atc-net/atc-dataplatform/pull/57#discussion_r780823779

File: examples/etl/multiple_transform_example.py (hunk 8 above)
Comment (mrmasterplan) [reply examples/etl/multiple_transform_example.py:42]: https://github.com/atc-net/atc-dataplatform/pull/57#discussion_r780823779

File: examples/etl/multiple_input_example.py (hunk 10 above)
Comment (mrmasterplan) [reply examples/etl/multiple_input_example.py:52]: I deliberately made them distinctly different to highlight the fact that the key was set manually to a new value.

File: src/atc/etl/transformer.py [hunk 13]
```diff
@@ -1,49 +1,30 @@
...
-        self.inner_transformers = inner_transformers
+    In regards to the etl step, the transformer CONSUMES the inputs
+    and ADDs the result of its transformation stage.
+    """
 
-    def get_transformers(self) -> List[Transformer]:
-        return self.inner_transformers
+    def etl(self, inputs: dataset_group) -> dataset_group:
+        if len(inputs) == 1:
+            for v in inputs.values():
```
Comment (JeppeBlixen) [inline src/atc/etl/transformer.py:20]: A bit wired to have a for loop for dict with len = 1?
The same could be written this way: next(iter(d.values()))

**Review by JeppeBlixen — APPROVED** [review]
I think it looks super great! I have also updated to fixed dependencies, so no worries for merge.

I have added some comments.

File: src/atc/etl/loader.py [hunk 14]
```diff
@@ -1,24 +1,31 @@
...
+    and does not consume or change it.
+    """
 
-class DelegatingLoader(Loader):
-    def __init__(self, inner_loaders: List[Loader]):
-        super().__init__()
-        self.inner_loaders = inner_loaders
+    def etl(self, inputs: dataset_group) -> dataset_group:
+        if len(inputs) == 1:
+            for k, v in inputs.items():
```
[the remark this answers is not in the document]
Comment (mrmasterplan) [reply src/atc/etl/loader.py:19]: Im not sure what is cleaner, I could also have gone with `list(inputs.values())[0]` I like your suggestion. let's go with that.

File: src/atc/etl/transformer.py [hunk 15]
```diff
@@ -1,49 +1,30 @@
...
-    def process(self, df: DataFrame) -> DataFrame:
-        return df
-
-
-class MultiInputTransformer:
-    @abstractmethod
-    def process_many(self, dataset: Dict[str, DataFrame]) -> DataFrame:
-        pass
 
+class Transformer(EtlBase):
```
[the remark this answers is not in the document]
Comment (mrmasterplan) [reply src/atc/etl/transformer.py:8]: I had considered it. Now I will do it.

**Review by christianhelle — APPROVED** [review]

**Pull request merged**